main.cpp

```diff
@@ -173,24 +240,63 @@ int main (int argc, char** argv)
 			current_handle = current_handle->next;
 		}
 
-
-		if (needrefresh)
-		{
-			if ((!DEBUG)&&(!tracemode))
+		time_t const now = ::time(NULL);
+ 		if( last_refresh_time + refreshdelay <= now )
+ 		{
+ 			last_refresh_time = now;
+ 			if ((!DEBUG)&&(!tracemode))
 			{
 				// handle user input
 				ui_tick();
 			}
 			do_refresh();
-			needrefresh = false;
 		}
-
-		// If no packets were read at all this iteration, pause to prevent 100%
-		// CPU utilisation;
+ 
+		//if not packets, do a select() until next packet
 		if (!packets_read)
 		{
-			usleep(100);
+			if( pc_loop_use_select )
+			{
+				FD_ZERO(&pc_loop_fd_set);
+				int nfds = 0;
+				for(std::vector<int>::const_iterator it=pc_loop_fd_list.begin();
+					it != pc_loop_fd_list.end(); ++it)
+				{
+					int const fd = *it;
+					nfds = std::max(nfds, *it + 1);
+					FD_SET(fd, &pc_loop_fd_set);
+				}
+				timeval timeout = {refreshdelay, 0};
+				if( select(nfds, &pc_loop_fd_set, 0, 0, &timeout) == -1 )
+				{
+					//this happens on system signal
+					continue;
+				}
+				if( FD_ISSET(self_pipe.first, &pc_loop_fd_set) )
+				{
+					std::cout << "exited by select\n";
+					//exit the loop
+					break;
+				}
+			}
+			else
+			{
+				// If select() not possible, pause to prevent 100%
+				// Pause 10 milliseconds
+				usleep(1000);
+			}
 		}
 	}
+	
+	//close file descriptors
+	for(std::vector<int>::const_iterator it=pc_loop_fd_list.begin();
+		it != pc_loop_fd_list.end(); ++it)
+	{
+		close(*it);
+	}
+	
+	procclean();
```

Comment:
Taking this out of the callback and moving it to right after the main loop is much nicer :+1: